internal/provider/provider.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	return &Provider{
+		omniEndpoint:          omniEndpoint,
+		omniServiceAccountKey: omniServiceAccountKey,
+		imageFactoryPXEURL:    imageFactoryPXEURL,
+		ipxeServerPort:        ipxeServerPort,
+		clear:                 clearState,
+		provisioner:           qemuProvisioner,
+		logger:                logger,
+	}, nil
+}
+
+// Run runs the provider.
+func (provider *Provider) Run(ctx context.Context) (runErr error) {
+	omniClient, err := provider.omniClient()
+	if err != nil {
+		return err
+	}
+
+	omniState := omniClient.Omni().State()
+
+	if provider.clear {
+		if err = provider.clearState(ctx, omniState); err != nil {
+			return fmt.Errorf("failed to clear provider state: %w", err)
+		}
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	eg.Go(func() error {
+		return provider.watchSysVersion(ctx, omniClient.Omni().State())
+	})
+
+	eg.Go(func() error {
+		server := ipxe.NewServer(omniState, provider.imageFactoryPXEURL, provider.ipxeServerPort, provider.logger)
+
+		ipxeServerErr := server.Run(ctx)
+		if ipxeServerErr != nil {
+			provider.logger.Error("failed to start iPXE server", zap.Error(ipxeServerErr))
+		}
+
+		return ipxeServerErr
+	})
+
+	eg.Go(func() error {
+		provisionErr := provider.provisioner.Run(ctx)
+		if provisionErr != nil {
+			provider.logger.Error("failed to provision machines", zap.Error(provisionErr))
+		}
+
+		return provisionErr
+	})
+
+	defer func() {
+		if closeErr := omniClient.Close(); closeErr != nil {
+			runErr = errors.Join(runErr, fmt.Errorf("failed to close Omni client: %w", closeErr))
+		}
+	}()
+
+	if err = protobuf.RegisterResource(resources.QemuMachineType, &resources.QemuMachine{}); err != nil {
+		return fmt.Errorf("failed to register resource: %w", err)
+	}
+
+	if err = protobuf.RegisterResource(resources.QemuMachineAllocationType, &resources.QemuMachineAllocation{}); err != nil {
+		return fmt.Errorf("failed to register resource: %w", err)
+	}
+
+	cosiRuntime, err := runtime.NewRuntime(omniState, provider.logger)
+	if err != nil {
+		return fmt.Errorf("failed to create runtime: %w", err)
+	}
+
+	if err = cosiRuntime.RegisterQController(controller.NewMachineRequestStatusController(provider.provisioner)); err != nil {
+		return fmt.Errorf("failed to register controller: %w", err)
+	}
+
+	if err = cosiRuntime.Run(ctx); err != nil {
+		return fmt.Errorf("failed to run runtime: %w", err)
+	}
+
+	return eg.Wait()
+}
+
+// watchSysVersion serves as a health check - if the connection to the Omni API is lost (e.g., due to an Omni restart),
+// this will return an error and the provider will crash.
+func (provider *Provider) watchSysVersion(ctx context.Context, st state.State) error {
+	eventCh := make(chan state.Event)
+
+	if err := st.Watch(ctx, system.NewSysVersion(omniresources.EphemeralNamespace, system.SysVersionID).Metadata(), eventCh); err != nil {
```

Comment:
I wonder if we should add a proper health endpoint.

Reply:
Yep, that's a good idea. Here the problem was, when Omni restarted, the `state` I had over the client API was broken, but the runtime wasn't aware of it. I couldn't find a way to make it aware of it (using GRPC options etc.). Maybe there is one, but I couldn't find any.

So not sure if it's an Omni problem - maybe a COSI one?